### ISSUE TITLE
(azure-functions-maven-plugin) Deployment fails to restricted storage account

### ISSUE THREAD

**Issue by melzayet** [Microsoft/azure-maven-plugins#2058]
### Plugin name and version
azure-functions-maven-plugin 1.17.0

### Plugin configuration in your `pom.xml`
```
               <configuration>
                    <resourceGroup>${functionResourceGroup}</resourceGroup>
                    <appName>${functionAppName}</appName>
                    <region>${functionAppRegion}</region>                                   
                    <runtime>
                        <os>linux</os>
                        <javaVersion>11</javaVersion>
                    </runtime>
                    <appSettings>
                        <!-- Run Azure Function from package file by default -->
                        <property>
                            <name>FUNCTIONS_EXTENSION_VERSION</name>
                            <value>~4</value>
                        </property>
                        <property>
                            <name>FUNCTIONS_WORKER_RUNTIME</name>
                            <value>java</value>
                        </property>
                    </appSettings>
                </configuration>
```
### Expected behavior
Deploy successfully to restricted storage account described [here](https://docs.microsoft.com/en-us/azure/azure-functions/configure-networking-how-to#restrict-your-storage-account-to-a-virtual-network) :
1. Azure Premium Function (linux) using storage account with firewall whitelisting function's subnet
2. Azure Function using regional VNET integration
3. Azure Function has WEBSITE_CONTENTOVERVNET=1

### Actual behavior
Received the following error:
```
deploy to Azure Function App with resource creation or updating: AzureToolkitRuntimeException: update Function app (xxxxxx): Status code 400, "{"Code":"BadRequest","Message":"Storage access failed. Storage volume is currently in R/O mode","Target":null,"Details":[{"Message":"Storage access failed. Storage volume is currently in R/O mode"},{"Code":"BadRequest"},{"ErrorEntity":{"ExtendedCode":"04234","MessageTemplate":"Storage access failed. {0}","Parameters":["Storage volume is currently in R/O mode"],"Code":"BadRequest","Message":"Storage access failed. Storage volume is currently in R/O mode"}}],"Innererror":null}" ->
```

The deployment works only if firewall is disabled or if using Azure CLI for Zip push deployment of MVN target subfolder:
```
az functionapp deployment source config-zip -g <resource-group> -n <function-name> --src <zip-file>
```

### Steps to reproduce the problem
1. Follow this sample to build locally: https://github.com/Azure-Samples/hello-spring-function-azure
2. Set the azure configuration described in "Expected Behaviour". Then deploy to azure using standard MVN deploy command: ./mvnw azure-functions:deploy

**Comment by Flanker32** [comment]
@melzayet Thanks for your report, for linux premium function, plugin will use [run from package](https://docs.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package) for deployment by default, which will upload artifact to storage and set its sas token to app settings, but it should be blocked by the firewall.

In this case, you may use ZIP deployment with `<deploymentType>ZIP</deploymentType>`, which should work same as azure cli.  Please refer [here](https://github.com/microsoft/azure-maven-plugins/wiki/Azure-Functions:-Configuration-Details#supported-deployment-type) for details.

**Comment by melzayet** [comment]
@Flanker32 thanks a lot for the clarification. So if the Maven plugin is running in a build agent running in a network with access to storage account, this would allow it pass through the firewall right?
In other words, does the Maven plugin do a direct call to storage account to upload the zip file?

**Comment by melzayet** [comment]
@Flanker32 To follow up on the recommendation above of changing deployment type to ZIP, the same error still occurs. Also happens with RUN_FROM_ZIP. We doubled checked and the deploymentType configuration was correctly placed in the pom.xml. Is there any chance these type of ZIP deployments would require storage account access?

**Comment by Flanker32** [comment]
> @Flanker32 thanks a lot for the clarification. So if the Maven plugin is running in a build agent running in a network with access to storage account, this would allow it pass through the firewall right? In other words, does the Maven plugin do a direct call to storage account to upload the zip file?

For RUN_FROM_BLOB (default for Linux premium and consumption), yes, we will call storage api to upload the artifacts

**Comment by Flanker32** [comment]
> @Flanker32 To follow up on the recommendation above of changing deployment type to ZIP, the same error still occurs. Also happens with RUN_FROM_ZIP. We doubled checked and the deploymentType configuration was correctly placed in the pom.xml. Is there any chance these type of ZIP deployments would require storage account access?

Could you please help share the full error stack with `mvn -X`? For there two ZIP deployment, we just call kudu zipdeploy which should not have storage call.

**Comment by melzayet** [comment]
@Flanker32 Thanks. I have attached the full error stack as mentioned. I just have an update here. While this deploymentType did not work with an EP1 plan with Azure File share being attached, as soon as we moved into a Dedicated Service Plan (P1v2), the plugin deployment worked straight away with same pom.xml

In the successful deployment that used Dedicated Service Plan, I could see the logs showing ZIP deploy to SCM endpoint and the setting of RUN_FROM_PACKAGE. Yet that was not happening in the attached logs that come from an elastic plan EP1 deployment

[debugfullstack.txt](https://github.com/microsoft/azure-maven-plugins/files/9000894/debugfullstack.txt)

**Comment by Flanker32** [comment]
@melzayet Thanks for your help and sorry for the late reply! It seems that exception occurred when plugin trying to update the function app rather than the deployment. It may related to https://github.com/microsoft/azure-maven-plugins/issues/2052 that maven plugin will do some extra update during deployment. The issue has been resolved in plugin v1.19.0, could you please try the new version and see whether the issue could be reproduced?

**Comment by melzayet** [comment]
@Flanker32, my apologies for such a late reply. I tested this today with v1.19.0 and works perfectly with Zip deployment type into EP1. Looks like #2052 did solve that. Many thanks for you help on this issue. I will close it now
